interpreter/interpreter.go

```diff
@@ -151,6 +156,230 @@ func CompileRegexConstants(regexOptimizations ...*RegexOptimization) Interpretab
 	return decRegexOptimizer(regexOptimizations...)
 }
 
+type typeVerifier func(int64, ...*types.TypeValue) (bool, error)
+
+// InterpolateFormattedString checks the syntax and cardinality of any string.format calls present in the expression and reports
+// any errors at compile time.
+func InterpolateFormattedString(enableCheckArgTypes bool, verifier typeVerifier) InterpretableDecorator {
+	return func(inter Interpretable) (Interpretable, error) {
+		call, ok := inter.(InterpretableCall)
+		if !ok {
+			return inter, nil
+		}
+		if call.OverloadID() != "string_format" {
+			return inter, nil
+		}
+		args := call.Args()
+		if len(args) != 2 {
+			return nil, fmt.Errorf("wrong number of arguments to string.format (expected 2, got %d)", len(args))
+		}
+		fmtStrInter, ok := args[0].(InterpretableConst)
+		if !ok {
+			return inter, nil
+		}
+		var fmtArgsInter InterpretableConstructor
+		fmtArgsInter, ok = args[1].(InterpretableConstructor)
+		if !ok {
+			// check for an evalWatch wrapper, since evalWatch implements
```

Comment:
This seems like a short-coming of the watcher code. Could you instead create an `evalWatchConstructor` case in `decorators.go` that exposes the right interface for `InterpretableConstructor`? If you do that, then you can move the whole function and associated supporting code into `ext.Strings()` 

Moving the implementation up also allows you to use the cel/decls.go objects in your code rather than have to do multiple conversions between runtime type, expr protobuf type, and cel type.

Reply:
Wouldn't there be an import cycle even after adding an `evalWatchConstructor` if we moved `InterpolateFormattedString` into `ext/strings.go`? The extension library already imports `"github.com/google/cel-go/cel"`, and if we moved `InterpolateFormattedString`, the flag check would read something like:

```go
        // Enable compile-time checking of syntax/cardinality for string.format calls.
	if p.evalOpts&OptCheckStringFormat == OptCheckStringFormat {
		...
		decorators = append(decorators, strings.InterpolateFormattedString(enableCheckArgTypes, isValidType))
	}
```

Definitely agree on implementing `evalWatchConstructor` though, I'll do that shortly.

Reply:
Ideally, you'd be able to return the decorator as part of the `stringLib.ProgramOptions()`; however, you need the ast reference, so you have to defer it until here. This is something I'll have to address later then, but let me see if I can think of any convenient ways of approaching this sooner.

Reply:
Think this is fine to mark as resolved since `evalWatchConstructor` is implemented. I'll track moving `InterpolateFormattedString` in  https://github.com/google/cel-go/pull/645#discussion_r1122454806.